File: Computer Programming/C++/Recursion and Dynamic Programming/Combination Sum IV.cpp
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) 
    {
        int n = nums.size();    

        vector<vector<int>> vis(n+1,vector<int>(target+1,-1));
        int ans = function(vis,nums,target,n-1);
        return ans;
    }

    int function(vector<vector<int>>&vis,vector<int>&nums,int target, int index)
    {
        int n = nums.size();
        if(target == 0)
        {
            return 1;
        }
        if(target < 0)
        {
            return 0;
        }

        if(index < 0)
        {
            return 0;
        }
        if(vis[index][target] != -1)
        {
            return vis[index][target];
        }

        int pick = function(vis,nums,target - nums[index],n- 1);
        int notpick = function(vis,nums,target,index-1);

        return vis[index][target] = pick + notpick;
    }
};

File: Computer Programming/Python/Dynamic Programming/Combination Sum IV.py
class Solution(object):
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        
        n = len(nums)

        dp = []
        for i in range(n):
            row = []
            for j in range(target+1):
                row.append(-1)
            dp.append(row)

        ans = self.function(dp,nums,n-1,target)
        return ans

    def function(self,dp,nums,index,target):
        if target == 0:
            return 1
        if target < 0:
            return 0
        if index < 0:
            return 0

        if dp[index][target] != -1:
            return dp[index][target]

        n = len(nums)
        pick = self.function(dp,nums,n-1,target - nums[index])
        notpick = self.function(dp,nums,index-1,target)
        
        dp[index][target] = pick + notpick

        return dp[index][target]


         
